### PR TITLE
Apply batched branch modifications efficiently, but in the correct order

### LINKED ISSUE #1552: stepManyAt0 should reflect the order of input actions
There was a possibly-unwritten expectation of `stepManyAt0` that it should produce the same resulting namespace as if the actions had been executed sequentially.

Breaking this resulted in #1430 / #1495 (and possibly undiscovered others).

I suspect this was a regression in #1225/#1333; e.g. working from the children to the root causes these actions to happen out of the expected order:

```
MoveBranchI Nothing dest -> do
  b <- use root
  stepManyAt [ (Path.empty, const Branch.empty0)
             , BranchUtil.makeSetBranch (resolveSplit' dest) b ]
  success
```

and the root gets wiped as the last step instead of the first step.

Maybe we can't generally consolidate everything into a single action, especially if individual actions are allowed to change things deeper than their rooted path, but hopefully we can do something to minimize the number.

### PR DESCRIPTION
**Pull request by ChrisPenner**
Efficiently apply many actions across many namespace children WITHOUT breaking dependencies between actions.

The current code applies all actions at the current node, then all actions at children, but it's possible for the caller to want some actions to happen on children first, then some at the current branch, then more at the children, etc.

closes #1552 

## Overview

This should prevent accidental namespace deletions, and similar logic bugs when batching actions together.

## Implementation notes

* Rephrases all of `step`, `stepM`, `stepManyAt`, `stepManyAtM`, etc. to all go through `batchUpdates` for consistency
* Rewrites our batch update logic (previously called `stepManyAt0`; now it batches up and applies non-conflicting updates for child branches as well as it can, but preserves semantically correct ordering in cases where actions might depend-on or affect each other.
* The new `stepManyAt` functions perform all updates, then FLATTENS them all into a single cons onto the starting branch. This flattens more aggressively than the previous implementation which would add one step in the root, but possibly many steps over the child branches (if an action performed it's own manipulation over history).
* This is implemented in terms of `consBranch` which performs the flattening. It's distinct from a squash because it doesn't perform any LCA checks or merge histories, since we don't want any of that in `stepManyAt`.
* Fixes broken GitSync transcript saving (the work-dir it's run in has changed)
* Adds regression tests for moving a namespace to a child of its current namespace (the motivation for this PR) AND records the current behaviour of merging with respect to child branches. We may wish to change how squash merges are handled on children in the future.

## Test coverage

- [x] Added a regression transcript for the namespace deletion issue.
- [x] Added a transcript recording current behaviour of merges and squash merges over child branches.

### CODE REVIEW

**Review by pchiusano — APPROVED** [review]
Nice! Glad to get this finally fixed.

**Comment by pchiusano** [comment]
CI is failing due to transcript differences. I'd refresh transcripts and convince yourself the differences are all expected. :)

File: unison-src/transcripts/squash.output.md
Comment (ChrisPenner) [inline unison-src/transcripts/squash.output.md]: Interesting that this was removed; not sure what's up here but I'll look into it.

Either I broke something or fixed something haha.

File: unison-src/transcripts/squash.output.md
Comment (ChrisPenner) [inline unison-src/transcripts/squash.output.md]: Same here, somewhat curious behaviour.

File: unison-src/transcripts/reflog.output.md
Comment (ChrisPenner) [inline unison-src/transcripts/reflog.output.md]: Not sure whether all these hashes are expected to change, I suppose it's possible that re-ordering operations could result in a new state, or maybe just the ordering of parents changed somehow due to re-ordering of action, but I'll have to look into it to see if it's equivalent, or if I broke or fixed something 🤔

I think it's quite likely that what used to be a single history step on a sub-branch has now been broken into two since there's now an action occurring over the parent in-between.

File: parser-typechecker/src/Unison/Codebase/Branch.hs
Comment (ChrisPenner) [inline parser-typechecker/src/Unison/Codebase/Branch.hs:623]: I notice this doesn't use `step`; should it?

**Comment by aryairani** [comment]
Sorry for not mentioning it sooner, but lets merge #2572 before we merge this PR because of

> I totally solved the problem of not being able to force delete namespaces! You just take advantage of https://github.com/unisonweb/unison/issues/1430

File: unison-src/transcripts/reflog.output.md
Comment (aryairani) [reply unison-src/transcripts/reflog.output.md]: Let's walk through the algorithm together before merging — we do want a multiple history steps on the sub-branches to be merged into one, with a squash step after all the actions have been run but before we finish and return to the input loop.

Comment (aryairani) [reply unison-src/transcripts/reflog.output.md]: I could believe that it's appropriate for some of these hashes to change though, we should dig into it though and see what the actual difference between the two histories is.

**Review by aryairani — COMMENTED** [review]
Testing: Does this count as a review?
Answer: Yes. It adds a 💬 icon as my review status, and removes it from my list of PRs waiting for review. I guess you'd have to "re-request review" for it to show up again.

I left some comments inline, but they didn't get grouped here.

File: unison-src/transcripts/reflog.output.md
Comment (ChrisPenner) [reply unison-src/transcripts/reflog.output.md]: Ah I see, so I should make sure I've run all changes (even from the parent branch) before committing the change to a new causal.

At a glance I think the old code might be making a mistake there too, since the action over the "root" branch might still make changes to the children after they've "stepped" from their own changes, but I should be able to fix up both of those (although it might still result in new hashes if I can get it making exactly one history step haha)

Comment (aryairani) [reply unison-src/transcripts/reflog.output.md]: Yeah the existing code definitely has a bug that caused #1430 / #1495.

**Comment by aryairani** [comment]
@ChrisPenner When you get a chance, would you post an update about the current state of what's currently in the PR vs what's left to do on this one?

**Comment by ChrisPenner** [comment]
@aryairani  I'm waiting for reviews on #2652 so I can merge that first, and I need to do a few more tests to confirm that the changed transcript hashes all make sense. Changed hashes are expected since the new squash reduces the number of changes even further than it used to, but just have to do a sanity check to ensure it's all reasonable 😄 

I don't expect any major changes here though, so you and @mitchellwrosen  can review when whenever 😄

File: parser-typechecker/src/Unison/Codebase/Branch.hs
Comment (ChrisPenner) [inline parser-typechecker/src/Unison/Codebase/Branch.hs:513]: I rewrote all of the pure variants in terms of their monadic counterparts using `Identity`; it felt safer to avoid the duplication wherever possible.

File: parser-typechecker/src/Unison/Codebase/Branch.hs
Comment (ChrisPenner) [inline parser-typechecker/src/Unison/Codebase/Branch.hs:632]: I renamed `stepManyAt0(M)` to `batchUpdates` since it doesn't actually do any stepping anymore.

File: parser-typechecker/src/Unison/Codebase/Branch.hs
Comment (ChrisPenner) [inline parser-typechecker/src/Unison/Codebase/Branch.hs:689]: Note here that we no longer `step` the children, we update the head as-is, it's not necessary to step and would add extra history that would just get squashed down anyways.

File: parser-typechecker/src/Unison/Codebase/Causal.hs
Comment (ChrisPenner) [inline parser-typechecker/src/Unison/Codebase/Causal.hs:101]: The previous lens was a bit unsafe since you'd end up with a causal whose hash doesn't match its contents; which caused a bug that I had to track down to here 😓

File: parser-typechecker/src/Unison/Codebase/Causal.hs
Comment (ChrisPenner) [inline parser-typechecker/src/Unison/Codebase/Causal.hs]: Open to ideas for a better name here

File: unison-src/transcripts/merges.output.md
Comment (ChrisPenner) [inline unison-src/transcripts/merges.output.md]: I believe these changed hashes are due to removing redundant steps on child branches in stepManyAt.

File: unison-src/transcripts/squash.output.md
Comment (ChrisPenner) [reply unison-src/transcripts/squash.output.md]: I think we're now just always adding a new cons for the merge regardless of whether there are no changes; I wonder if a child is being updated strangely causing a new hash and therefor not "identical" branches? I should double-check the logic in and around consdistinct to see if it's basing distinctness in any way on branch histories instead of just branch contents.

File: parser-typechecker/src/Unison/Codebase/Branch.hs
Comment (aryairani) [reply parser-typechecker/src/Unison/Codebase/Branch.hs:513]: Ok.

File: parser-typechecker/src/Unison/Codebase/Branch.hs
Comment (aryairani) [reply parser-typechecker/src/Unison/Codebase/Branch.hs:689]: Do the hashes get updated correctly in the end, or do we miss something by altering the head? If we don't `step`, then we need to be aware of any algorithms that might short-circuit incorrectly based on the cached head hash.

File: parser-typechecker/src/Unison/Codebase/Branch.hs
Comment (aryairani) [inline parser-typechecker/src/Unison/Codebase/Branch.hs]: A lot of these haddocks say a function does something efficiently, but what are they referring to?

File: parser-typechecker/src/Unison/Codebase/Branch.hs
Comment (aryairani) [reply parser-typechecker/src/Unison/Codebase/Branch.hs:689]: Ok, I see you discovered that.

File: parser-typechecker/src/Unison/Codebase/Causal.hs
Comment (aryairani) [reply parser-typechecker/src/Unison/Codebase/Causal.hs]: Maybe `mergeNode`?  Maybe this should go into `Causal.Type` as `merge` (although I don't remember if Causal.Type exists here yet, or if it's only in #2635, I think that's the case).  Or just name it whatever for now and then open a new issue to sort out all the "merge" name things.

File: unison-src/transcripts/namespace-deletion-regression.md
Comment (aryairani) [inline unison-src/transcripts/namespace-deletion-regression.md]: maybe `ls` `Nat` and `Nat.operators` too.  In any case, the two `ls` here show the same results before and after the `move.namespace` operation, so i'm not sure if that shows that it's worked correctly.

File: unison-src/transcripts/squash.output.md
Comment (aryairani) [reply unison-src/transcripts/squash.output.md]: Is this a question to self or to me or someone else? :)

File: parser-typechecker/src/Unison/Codebase/Branch.hs
Comment (aryairani) [inline parser-typechecker/src/Unison/Codebase/Branch.hs]: Also, what does this mean? :)

File: parser-typechecker/src/Unison/Codebase/Branch.hs
Comment (aryairani) [reply parser-typechecker/src/Unison/Codebase/Branch.hs:632]: Doesn't it still do one step for the children?  I see it calls `stepChildren`, but there's a PR comment there that says it doesn't step the children, and I'm confused.  What does this function do? :)

File: parser-typechecker/src/Unison/Codebase/Causal.hs
Comment (aryairani) [reply parser-typechecker/src/Unison/Codebase/Causal.hs:101]: Yes, sorry for not catching it at the time.

**Review by aryairani — COMMENTED** [review]
I think this might be fine? But I left many questions. 😅

**Comment by aryairani** [comment]
> @ChrisPenner When you get a chance, would you post an update about the current state of what's currently in the PR vs what's left to do on this one?

> @aryairani I'm waiting for reviews on #2652 so I can merge that first, and I need to do a few more tests to confirm that the changed transcript hashes all make sense. Changed hashes are expected since the new squash reduces the number of changes even further than it used to, but just have to do a sanity check to ensure it's all reasonable 😄
> 
> I don't expect any major changes here though, so you and @mitchellwrosen can review when whenever 😄

@ChrisPenner Ah ok, that makes sense.  But for my question, I meant: since we'd worked through a few different iterations of the implementation for this one (reordering the actions bottom up but some of the higher ones might clobber the lower ones; switching to a symbolic representation for branch updates; squashing the root node but not necessarily all child nodes; squashing all child nodes), I just wanted a quick confirmation of what algorithm the PR currently implements before diving into the implementation code.

The PR description currently is pretty high level -- for future PRs, anything in the description that can orients me as to what I'm looking at and what I should be looking for will be super helpful in streamlining review. You can call out renames or any other changes that aren't directly related to the title, so the reviewer isn't surprised midstream. And then because merging is all-or-nothing, where practical, avoid mixing possibly-controversial changes with uncontroversial ones, or with other possibly-controversial changes. (That's just a general guideline, I didn't see anything particularly controversial here!)

File: parser-typechecker/src/Unison/Codebase/Branch.hs
Comment (ChrisPenner) [reply parser-typechecker/src/Unison/Codebase/Branch.hs:632]: It used to step the children, I removed the actual step, now it updates the children in place. I'll rename `stepChildren` 👍🏼

File: unison-src/transcripts/squash.output.md
Comment (ChrisPenner) [reply unison-src/transcripts/squash.output.md]: This is me keeping track of my musings so that I remember what I was thinking when I get back into things on Monday 😓  😛

File: unison-src/transcripts/namespace-deletion-regression.md
Comment (ChrisPenner) [reply unison-src/transcripts/namespace-deletion-regression.md]: Good call 👍🏼

File: parser-typechecker/src/Unison/Codebase/Branch.hs
Comment (ChrisPenner) [inline parser-typechecker/src/Unison/Codebase/Branch.hs:187]: Killed some TH here (the splices were causing issues)

File: parser-typechecker/src/Unison/Codebase/Branch.hs
Comment (ChrisPenner) [inline parser-typechecker/src/Unison/Codebase/Branch.hs:364]: Previously this always appended onto an `empty`; but according to a [slack convo](https://unisoncomputing.slack.com/archives/CF8D3DCP3/p1637682445316200) we want to _start_ the branch with the first changes rather than having an empty branch there.

Comment (aryairani) [reply parser-typechecker/src/Unison/Codebase/Branch.hs:364]: Ah, right — though unintuitively, `Branch.cons x empty` was special cased (at whatever time we realized we wanted to not ever see a `cons x empty`) to equal `Branch.one x`

File: parser-typechecker/src/Unison/Codebase/Branch.hs
Comment (aryairani) [inline parser-typechecker/src/Unison/Codebase/Branch.hs]: I didn't understand the "Note"; was it meant to say it doesn't _squash_ history on child branches? Or does not stepping mean that it does squash them?

**Comment by ChrisPenner** [comment]
This is now acting as expected in transcripts and is ready for a final review and merge 😄

File: parser-typechecker/src/Unison/Codebase/Causal.hs
Comment (aryairani) [inline parser-typechecker/src/Unison/Codebase/Causal.hs:221]: `children` was certainly the wrong name.  WDYT about `tails`, although I guess that is currently a field name in the `Merge` case...

File: unison-src/transcripts/child-namespace-history-merge.md
Comment (aryairani) [inline unison-src/transcripts/child-namespace-history-merge.md]:
```suggestion
```unison:hide
```

This will hide the type-checking output, which is basically echoed verbatim in the `add` step.  Don't hide it if you want to show something about the type-checking error though, or don't plan to `add` immediately after and want to see the types printed.

File: unison-src/transcripts/child-namespace-history-merge.md
Comment (aryairani) [inline unison-src/transcripts/child-namespace-history-merge.md]:
```suggestion
```unison:hide
```

File: unison-src/transcripts/child-namespace-history-merge.md
Comment (aryairani) [inline unison-src/transcripts/child-namespace-history-merge.md]:
```suggestion
```unison:hide
```

File: parser-typechecker/src/Unison/Codebase/Branch.hs
Comment (aryairani) [inline parser-typechecker/src/Unison/Codebase/Branch.hs]: What does "efficiently" refer to in the title and various comments?  Does it just mean aggregated into a single causal step?

**Review by aryairani — COMMENTED** [review]
This feels very close.  Here are some comments I forgot to hit submit on earlier.

I'm kind of struggling to understand what's going on in `child-namespace-histotry-merge.md`; that's documenting some existing behavior about `merge.squash`, right?  Could you move that to a separate PR that we'll merge first, just to be clear on whether or not this PR changes the output of the transcript?

File: parser-typechecker/src/Unison/Codebase/Causal.hs
Comment (ChrisPenner) [reply parser-typechecker/src/Unison/Codebase/Causal.hs:221]: Isn't `predecessors` what you me and @mitchellwrosen  had decided on for Causals? I was pretty sure it was but we can revisit if you like. And yes, `tails` would conflict with the field name.

Mind if I ask what issue do you take with predecessors? 

> Predecessor: a thing that has been followed or replaced by another.

Seems like a good name to apply here 😄

File: parser-typechecker/src/Unison/Codebase/Branch.hs
Comment (ChrisPenner) [reply parser-typechecker/src/Unison/Codebase/Branch.hs]: The note is probably not relevant since all of the blessed ways of updating child branches will handle stepping properly anyways, I'll just remove the note.

**Comment by ChrisPenner** [comment]
@aryairani  I split off the transcript here: https://github.com/unisonweb/unison/pull/2688 😄

File: parser-typechecker/src/Unison/Codebase/Causal.hs
Comment (aryairani) [reply parser-typechecker/src/Unison/Codebase/Causal.hs:221]: Right, I don't really have a good answer.  I guess I was thinking predecessor/successor wasn't far enough from parent/child that was have in branches, but it kinda doesn't matter so let's just go with this.

File: unison-src/transcripts/merges.output.md
Comment (ChrisPenner) [inline unison-src/transcripts/merges.output.md:128]: I tracked down what caused this.

The stepManyAt on trunk loses history in child-namespaces when they're deleted. This new implementation keeps the history intact.

trunk:
```
.> history .feature1

  ☝️  The namespace .feature1 is empty.
```

This branch:
```
.> history .feature1

  Note: The most recent namespace hash is immediately below this
        message.
  
  ⊙ #gj7j96t88n
  
    - Deletes:
    
      y
  
  □ #aqh5c7qsod (start of history)
```

In this case, since nothing was deliberately squashed I think this branch's behaviour is preferable.
If you want to discard history, you should squash-merge instead.

File: unison-src/transcripts/merges.md
Comment (ChrisPenner) [inline unison-src/transcripts/merges.md:55]: Added a check on the child-branch's history. See comment on the output for the results.

File: unison-src/transcripts/merges.output.md
Comment (ChrisPenner) [inline unison-src/transcripts/merges.output.md:110]: The output on trunk is:

```
.> history .feature1 

☝️ The namespace .feature1 is empty.
```

File: unison-src/transcripts/merges.output.md
Comment (aryairani) [reply unison-src/transcripts/merges.output.md:128]: /cc @pchiusano I forget whether the old behavior was deliberate, coincidental, or known-undesirable; do you remember?

Comment (pchiusano) [reply unison-src/transcripts/merges.output.md:128]: Good catch. Old behavior is wrong, this PR looks more correct.

Deleting a subnamespace should not remove the history of that namespace from its parent. This is important for having merge do the right thing in more cases, since merge will then remember that the child namespace was deleted and not just resurrect the whole thing when there's an unrelated change at the same level.

**Review by pchiusano — APPROVED** [review]

**Comment by ChrisPenner** [comment]
@aryairani I'm going to assume Paul's response satisfies your concerns and get this merged 😄 
Let me know if this isn't the case and we can patch stuff up 😄

**Pull request merged**
